src/dune/ml_sources.ml

```diff
@@ -237,26 +247,19 @@ let make_lib_modules (d : _ Dir_with_dune.t) ~lookup_vlib ~(lib : Library.t)
         (* can't happen because this library is defined using the current stanza *)
         |> Option.value_exn
       in
-      (* diml: this [Result.ok_exn] means that if the user writes an invalid
-         [implements] field, we will get an error immediately even if the
-         library is not built. We should change this to carry the [Or_exn.t] a
-         bit longer. *)
-      let vlib =
-        Result.ok_exn
```

Comment:
@snowleopard this is the error that we are trying to delay. The issue is that the absence of a virtual library means that loading the rules in an entire directory will be broken.

The PR doesn't yet do that, but it goes in the right direction. The callers of `Dir_contents` should be turning `Error` values into `Build.fail`.

Reply:
It would be nice to add a test exposing this undesirable behaviour, so that this PR fixes it.

Reply:
(Otherwise, we will likely regress back at some point.)